pkg/common/glob/glob.go

```diff
@@ -0,0 +1,123 @@
+package common
+
+import (
+	"fmt"
+
+	"github.com/gobwas/glob"
+)
+
+// Filter is a generic filter for excluding and including globs (limited
+// regular expressions). Exclusion takes precedence if both include and exclude
+// lists are provided.
+type Filter struct {
+	exclude []glob.Glob
+	include []glob.Glob
+}
+
+type globFilterOpt func(*Filter) error
+
+// WithExcludeGlobs adds exclude globs to the filter.
+func WithExcludeGlobs(excludes ...string) globFilterOpt {
+	return func(f *Filter) error {
+		for _, exclude := range excludes {
+			g, err := glob.Compile(exclude)
+			if err != nil {
+				return err
+			}
+			f.exclude = append(f.exclude, g)
+		}
+		return nil
+	}
+}
+
+// WithIncludeGlobs adds include globs to the filter.
+func WithIncludeGlobs(includes ...string) globFilterOpt {
+	return func(f *Filter) error {
+		for _, include := range includes {
+			g, err := glob.Compile(include)
+			if err != nil {
+				return err
+			}
+			f.include = append(f.include, g)
+		}
+		return nil
+	}
+}
+
+// NewGlobFilter creates a new Filter with the provided options.
+func NewGlobFilter(opts ...globFilterOpt) (*Filter, error) {
+	filter := &Filter{}
+	for _, opt := range opts {
+		if err := opt(filter); err != nil {
+			return nil, err
+		}
+	}
+	return filter, nil
+}
+
+// Pass returns whether the object is not in the include list or in the exclude
+// list.
+func (f *Filter) Pass(object string) bool {
+	if f == nil {
+		return false
+	}
+	exclude, include := len(f.exclude), len(f.include)
+	if exclude == 0 && include == 0 {
+		return false
+	} else if exclude > 0 && include == 0 {
+		return f.passExclude(object)
+	} else if exclude == 0 && include > 0 {
+		return f.passInclude(object)
+	} else {
+		if pass, err := f.passExcludeInclude(object); err == nil {
+			return pass
+		}
+		// Ambiguous case. Should we skip? Should we not skip?
+		// Let's skip since the user indicated they want *some* form of
+		// filtering.
+		// TODO: Maybe this could be configurable.
+		return true
```

Comment:
Would adding a default case to the filter resolve this easily? It seems like it would take little effort but provide a lot more clarity. (Maybe make it three-state, where the third state is "return an error?")

Reply:
oh no i mean something like
```
type Filter struct {
	exclude []glob.Glob
	include []glob.Glob
        conflictBehavior bool
}
```
or
```
type Filter struct {
	exclude []glob.Glob
	include []glob.Glob
        conflictBehavior SomeNewTriStateType
}
```

Reply:
Oh I like it! If we do the tristate type, I think we would also have to return `(bool, error)` since that would be the only way to return an error on conflict.